### DISCUSSION TITLE
Network performance without relying on linux.

### DISCUSSION THREAD

**Discussion by deeprobin** [ubicloud/ubicloud#1321] [General]
Hello everyone,

I have just seen the [article](https://www.ubicloud.com/blog/improving-network-performance-with-linux-flowtables) about flowtables on your blog. 

Network performance is important, especially in the case of ubicloud.
Personally, I always use Linux as a rule. 

However, I once heard that FreeBSD and BSD systems in general have a very fast network stack that outperforms Linux.

Perhaps it would be relevant to shed some light on this? Especially for applications that are more network-driven.
Or is the aim of Ubicloud to rely exclusively on Linux systems?

See also this [thread](https://www.reddit.com/r/unix/comments/k40k85/how_is_freebsd_network_stack_still_faster_than/)

**Comment by fdr** [comment]
It would be interesting...but, we have a lot of reasons to be using Linux, probably most pertinent is the hypervisor, [Cloud Hypervisor](https://github.com/cloud-hypervisor/cloud-hypervisor/), which uses Linux KVM and MSHV backends. I suppose it would have to learn bhyve to make a comparable test. So it's pretty unlikely we can make a detour to try that out.

It is somewhat more likely, but with no immediate plan, that we'd consider a user-space stack like [VPP](https://github.com/FDio/vpp). If you were curious how, in general, network stacks could be more or less optimized, VPP, a very fast packet processor, provides good motivation. I suggest [media with Dave Barach](https://www.youtube.com/watch?v=9SLoVqfd1Xw) in it. I don't think anything prevents an OS kernel from implementing things more like VPP really, except (and this is a big "except") path dependency.
